source/dub/packagesupplier.d

```diff
@@ -266,27 +268,129 @@ class RegistryPackageSupplier : PackageSupplier {
 			.map!(j => SearchResult(j["name"].opt!string, j["description"].opt!string, j["version"].opt!string))
 			.array;
 	}
+}
+
+/**
+	Maven repository based package supplier.
+
+	This package supplier connects to a maven repository
+	to search for available packages.
+*/
+class MavenRegistryPackageSupplier : PackageSupplier {
+	private {
+		URL m_mavenUrl;
+		struct CacheEntry { Json data; SysTime cacheTime; }
+		CacheEntry[string] m_metadataCache;
+		Duration m_maxCacheTime;
+	}
 
-	private Json getBestPackage(string packageId, Dependency dep, bool pre_release)
+	this(URL mavenUrl)
 	{
-		Json md = getMetadata(packageId);
+		m_mavenUrl = mavenUrl;
+		m_maxCacheTime = 24.hours();
+	}
+
+	override @property string description() { return "maven repository at "~m_mavenUrl.toString(); }
+
+	Version[] getVersions(string package_id)
+	{
+		auto md = getMetadata(package_id);
 		if (md.type == Json.Type.null_)
-			return md;
-		Json best = null;
-		Version bestver;
+			return null;
+		Version[] ret;
 		foreach (json; md["versions"]) {
-			auto cur = Version(cast(string)json["version"]);
-			if (!dep.matches(cur)) continue;
-			if (best == null) best = json;
-			else if (pre_release) {
-				if (cur > bestver) best = json;
-			} else if (bestver.isPreRelease) {
-				if (!cur.isPreRelease || cur > bestver) best = json;
-			} else if (!cur.isPreRelease && cur > bestver) best = json;
-			bestver = Version(cast(string)best["version"]);
+			auto cur = Version(json["version"].get!string);
+			ret ~= cur;
+		}
+		ret.sort();
+		return ret;
+	}
+
+	void fetchPackage(NativePath path, string packageId, Dependency dep, bool pre_release)
+	{
+		auto md = getMetadata(packageId);
+		Json best = getBestPackage(md, packageId, dep, pre_release);
+		if (best.type == Json.Type.null_)
+			return;
+		auto vers = best["version"].get!string;
+		auto url = m_mavenUrl~NativePath("%s/%s/%s-%s.zip".format(packageId, vers, packageId, vers));
+		logDiagnostic("Downloading from '%s'", url);
+		foreach(i; 0..3) {
+			try{
+				download(url, path);
+				return;
+			}
+			catch(HTTPStatusException e) {
+				if (e.status == 404) throw e;
+				else {
+					logDebug("Failed to download package %s from %s (Attempt %s of 3)", packageId, url, i + 1);
+					continue;
+				}
+			}
+		}
+		throw new Exception("Failed to download package %s from %s".format(packageId, url));
+	}
+
+	Json fetchPackageRecipe(string packageId, Dependency dep, bool pre_release)
+	{
+		auto md = getMetadata(packageId);
+		return getBestPackage(md, packageId, dep, pre_release);
+	}
+
+	private Json getMetadata(string packageId)
+	{
+		import std.xml;
```

Comment:
not sure if it's a good idea to import this, vibe should really contain something for xml or at least html parsing

Reply:
I checked, unfortunately vibed doesn't contain anything regarding xml. It would maybe possible to parse the text via regex (Find all child "version" tags within parent tag "versions") or plain string functions. But as phobos contains a xml parsers which works quite well for this scenario, I am not 100% sure.